hazelcast/include/hazelcast/client/serialization/pimpl/compact.h

```diff
@@ -1407,6 +1407,13 @@ class HAZELCAST_API default_compact_writer
 
 struct HAZELCAST_API field_descriptor
 {
+    field_descriptor(
+        enum field_kind f = static_cast<enum field_kind>( -1 ),
```

Comment:
As I see, the enum values differ in C++ and Java APIs. In C++, it is one less than Java. For example, BOOLEAN is 1 in Java and 0 in C++. That makes NOT_AVAILABLE is missing in C++. I think that is why we used -1 as a default value. As I see from the other part of the code, there is a possibility to access arrays with this field_kind (field_operations::get(enum field_kind field_kind)). So maybe we it might be better to add an enumaration value with a positive value.

Reply:
Enum values are incorrect for this PR, I didn't correct those values because I was unaware of it at the time of writing this commit. Anyway, I think it is not in the scope of this PR, just added some unit tests. At replicate schema on write PR, [here](https://github.com/hazelcast/hazelcast-cpp-client/pull/1039/files#diff-90e0938a1ce08badb5a982647a1eee31484c2cec058fb26612cfc1afdefa206bR27) I corrected those enum values.

NOT_AVAILABLE has use when GenericRecord is introduced. Otherwise it is not used in Java which is not implemented yet in C++ too. So I don't think it needs to be added at this point. I plan to add it with GenericRecord implementation.

field_kind is initialized as -1 because any positive value is either vaiid, false-positive or out-of-bound. That's why I chose -1 to express that it is not initialized.

For https://github.com/hazelcast/hazelcast-cpp-client/pull/1039 i can make it 0.

Reply:
As it will be changed in the other PR, I think we can go on with current implementation.